lib/internal/modules/esm/get_source.js

```diff
@@ -5,28 +5,28 @@ const {
   decodeURIComponent,
 } = primordials;
 const { getOptionValue } = require('internal/options');
+const esmLoader = require('internal/process/esm_loader');
+
 // Do not eagerly grab .manifest, it may be in TDZ
 const policy = getOptionValue('--experimental-policy') ?
   require('internal/process/policy') :
   null;
 
 const { Buffer } = require('buffer');
 
-const fs = require('internal/fs/promises').exports;
 const { URL } = require('internal/url');
 const {
   ERR_INVALID_URL,
   ERR_INVALID_URL_SCHEME,
 } = require('internal/errors').codes;
-const readFileAsync = fs.readFile;
 
 const DATA_URL_PATTERN = /^[^/]+\/[^,;]+(?:[^,]*?)(;base64)?,([\s\S]*)$/;
 
 async function defaultGetSource(url, { format } = {}, defaultGetSource) {
   const parsed = new URL(url);
   let source;
   if (parsed.protocol === 'file:') {
-    source = await readFileAsync(parsed);
+    source = await esmLoader.getFileSystem().readFile(parsed);
```

Comment:
it would be preferable to have this passed in as a param instead of grabbing off of the application context `esmLoader` (there can be multiple Loaders).

Reply:
In which case does that happen? It could be a blocker for the `node:resolution` accessors I mentioned earlier, as it relies on loader utils being singletons.

If that's the case of the "loader loader" that @JakobJingleheimer mentions [there](https://github.com/nodejs/node/pull/41076#discussion_r765240614), since from what I understand they aren't used at the same time, perhaps the global access would be fine?

Reply:
I THINK their sequence is fully serial: the node-land esmLoaderNodeland finishes its work and disappears into the void, and then esmLoaderUserland steps in and starts its work.

Aside from Bradley's concern of bloating params, I think it would be better for this to be passed (and also makes it easier to test). One of our Loader team's side/wish-list projects is to make ESMLoader pure, and this goes against that. Not necessarily a show-stopper, but something to mention: If we do decide to make it global and preclude fully pure, we should do it consciously.